js/index.js

```diff
@@ -39,117 +44,187 @@ class Snake extends Grid {
             } else {
                 cellElement.classList.add(Snake.snakeBodyCssClass, Snake.snakeCssClass);
             }
         }
     }
 
+    #generateFood() {
+        do {
+            this.#food = {
+                cell: Math.floor(Math.random() * this.gridCount),
+                row: Math.floor(Math.random() * this.gridCount),
+            };
+        } while (
+            this.#snake.find((element) => {
+                return element.cell === this.#food.cell && element.row === this.#food.row;
+            })
+        );
+
+        const cellFood = this.#findByCoords(this.#food);
+        const image = new Image();
+        image.src = this.#foodSrcImage;
+        cellFood.append(image);
+    }
+
+    #checkHasFoodEaten() {
+        if (this.#snake[0].row === this.#food.row && this.#snake[0].cell === this.#food.cell) {
+            this.#score = ++this.#score;
+            this.#snake.push(this.#food);
+
+            const snakeContainer = document.querySelector(Snake.gridContainerCssSelector);
+
+            for (let i = 0; i < snakeContainer.children.length; i++) {
+                const element = snakeContainer.children[i];
+                element.innerHTML = '';
+            }
+
+            this.#generateFood();
+        }
+    }
+
+    #checkOnTailCrash() {
+        const snakeHead = this.#snake[0];
+
+        const snakeBiteItself =
+            this.#snake.filter((element) => {
+                return element.cell === snakeHead.cell && element.row === snakeHead.row;
+            }).length - 1;
+
+        if (snakeBiteItself >= 1) {
+            this.#end();
+        }
+    }
+
     #clear() {
-        let cells = this.find(`.${Snake.snakeCssClass}`, this.gridContainer);
-        
-        cells.forEach(cell => {
+        const cells = this.find(`.${Snake.snakeCssClass}`, this.gridContainer);
+        cells.forEach((cell) => {
             cell.className = Snake.snakeCellCssClass;
-            console.log('cell', cell.className);
         });
     }
-    
+
     #updateDirection(event) {
-        let key = event.key;
-        console.log(event);
+        const { key } = event;
+
+        if (key === 'ArrowLeft' && this.direction !== DR.RIGHT) this.direction = DR.LEFT;
+        else if (key === 'ArrowUp' && this.direction !== DR.DOWN) this.direction = DR.UP;
+        else if (key === 'ArrowRight' && this.direction !== DR.LEFT) this.direction = DR.RIGHT;
+        else if (key === 'ArrowDown' && this.direction !== DR.UP) this.direction = DR.DOWN;
+    }
+
+    #resetData() {
+        this.#snake = [];
+        this.#process = null;
+        this.#score = 0;
+        this.#speed = 0;
+        this.#food = null;
+        this.#mode = '';
+        this.direction = DR.LEFT;
+        this.#scoreContainer.lastChild.innerHTML = '0';
+
+        this.#controls.speed.removeAttribute('disabled');
+        this.#controls.mode.removeAttribute('disabled');
```

Comment:
you should put this code in separate method #enableControls, and use it here in the resetData()
```suggestion
        #enableControls() {
             this.#controls.speed.removeAttribute('disabled');
             this.#controls.mode.removeAttribute('disabled');
        }
         
```